python/dgl/contrib/sampling/sampler.py

```diff
@@ -434,6 +435,148 @@ def fetch(self, current_nodeflow_index):
         nflows = [NodeFlow(self.g, obj) for obj in nfobjs]
         return nflows
 
+
+class EdgeSampler(object):
+    '''Edge sampler for link prediction.
+
+    This samples edges from a given graph. The edges sampled for a batch are
+    placed in a subgraph before returning. In many link prediction tasks,
+    negative edges are required to train a model. A negative edge is constructed by
+    corrupting an existing edge in the graph. The current implementation
+    support two ways of corrupting an edge: corrupt the head node of
+    an edge (by randomly selecting a node as the head node), or corrupt
+    the tail node of an edge. When we corrupt the head node of an edge, we randomly
+    sample a node from the entire graph as the head node. It's possible the constructed
+    edge exists in the graph. By default, the implementation doesn't explicitly check
+    if the sampled negative edge exists in a graph. However, a user can exclude
+    positive edges from negative edges by specifying 'exclude_positive=True'.
+    When negative edges are created, a batch of negative edges are also placed
+    in a subgraph.
```

Comment:
So essentially this sampler
(1) randomly picks some edges in the graph, and put them in a subgraph (called "positive subgraph"),
(2) randomly picks some pairs of nodes in the graph, regardless of whether the pairs actually have connections in the graph, and put the (maybe-non-existent) edges in another subgraph (called "negative subgraph").

In other words, this is only useful for knowledge graph embedding methods like TransE.

Correct?

Reply:
not really. I think it's useful for any link prediction tasks.

Reply:
originally, I wanted to do neighbor sampling on end-points nodes as well. Then Xinjian suggested me to separate them to make it easier for users to understand.

Reply:
> not really. I think it's useful for any link prediction tasks.

Since we don't have NodeFlow for the nodes, I don't think it is useful _per se_ for models like GCMC and PinSage.

> originally, I wanted to do neighbor sampling on end-points nodes as well. Then Xinjian suggested me to separate them to make it easier for users to understand.

What he suggested I think is to generate seed nodes with a sampler, then generate NodeFlows from the first sampler's output.